### PR TITLE
Do not throw in PartitionedOutputBufferManager::updateBroadcastOutputBuffers()

### PR DESCRIPTION
**Pull request by karteekmurthys**
Replace getBuffer() with getBufferIfExists() in
PartitionedOutputBufferManager::updateBroadcastOutputBuffers() to avoid
throwing an exception if the buffer associated with the given task has been
deleted already.

See #3593 for more details.

### CODE REVIEW

File: velox/exec/PartitionedOutputBufferManager.cpp [hunk 1]
```diff
@@ -570,11 +570,16 @@ void PartitionedOutputBufferManager::initializeTask(
...
 }
 
-void PartitionedOutputBufferManager::updateBroadcastOutputBuffers(
+bool PartitionedOutputBufferManager::updateBroadcastOutputBuffers(
     const std::string& taskId,
     int numBuffers,
     bool noMoreBuffers) {
-  getBuffer(taskId)->updateBroadcastOutputBuffers(numBuffers, noMoreBuffers);
+  auto buffer = getBufferIfExists(taskId);
+  if (buffer) {
```
Comment (mbasmanova) [inline velox/exec/PartitionedOutputBufferManager.cpp:578]: nit: `if (auto buffer = getBufferIfExists(taskId))`

File: velox/exec/Task.cpp [hunk 2]
```diff
@@ -1002,7 +1002,7 @@ bool Task::isFinishedLocked() const {
   return (state_ == TaskState::kFinished);
 }
 
-void Task::updateBroadcastOutputBuffers(int numBuffers, bool noMoreBuffers) {
+bool Task::updateBroadcastOutputBuffers(int numBuffers, bool noMoreBuffers) {
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:1005]: I don't think this change is needed (or correct) and the return value doesn't appear to be used.

File: velox/exec/Task.cpp [hunk 3]
```diff
@@ -1488,7 +1488,7 @@ ContinueFuture Task::stateChangeFuture(uint64_t maxWaitMicros) {
   auto [promise, future] = makeVeloxContinuePromiseContract(
       fmt::format("Task::stateChangeFuture {}", taskId_));
   stateChangePromises_.emplace_back(std::move(promise));
-  if (maxWaitMicros > 0) {
+  if (maxWaitMicros) {
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:1491]: Unrelated change?

**Review by mbasmanova — COMMENTED** [review]
@karteekmurthys Thank you for following up on this issue. I don't think we need changes in Task.h,cpp, but the rest of the changes look good.

File: velox/exec/Task.cpp (hunk 2 above)
Comment (karteekmurthys) [reply velox/exec/Task.cpp:1005]: I am intending to use here @mbasmanova . If return value is false and task was aborted or cancelled, then we return from createOrUpdateTask with TaskInfo as return value.
https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/TaskManager.cpp#L339

Comment (mbasmanova) [reply velox/exec/Task.cpp:1005]: I see. Thank you for clarifying.

File: velox/exec/Task.cpp [hunk 4]
```diff
@@ -1013,15 +1013,15 @@ void Task::updateBroadcastOutputBuffers(int numBuffers, bool noMoreBuffers) {
     std::lock_guard<std::mutex> l(mutex_);
     if (noMoreBroadcastBuffers_) {
       // Ignore messages received after no-more-buffers message.
-      return;
+      return true;
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:1016]: Should we return false here since we are not updating the buffers?

File: velox/exec/Task.h [hunk 5]
```diff
@@ -194,7 +194,8 @@ class Task : public std::enable_shared_from_this<Task> {
   /// of buffers. No more calls are expected after the call with noMoreBuffers
   /// == true, but occasionally the caller might resend it, so calls
   /// received after a call with noMoreBuffers == true are ignored.
-  void updateBroadcastOutputBuffers(int numBuffers, bool noMoreBuffers);
+  /// @return bool Return true if the update was successful, else false.
```
Comment (mbasmanova) [inline velox/exec/Task.h:197]: Maybe clarify why the update may fail, e.g. the task execution fails or this calls come after a call with noMoreBuffers == true.

File: velox/exec/tests/PartitionedOutputBufferManagerTest.cpp [hunk 6]
```diff
@@ -428,3 +428,12 @@ TEST_F(PartitionedOutputBufferManagerTest, getDataOnFailedTask) {
         VELOX_UNREACHABLE();
       }));
 }
+
+TEST_F(PartitionedOutputBufferManagerTest, updateBrodcastBufferOnFailedTask) {
```
Comment (mbasmanova) [inline velox/exec/tests/PartitionedOutputBufferManagerTest.cpp:432]: It would be nice to add a test for the changes in Task.h,cpp.

File: velox/exec/Task.cpp (hunk 4 above)
Comment (karteekmurthys) [reply velox/exec/Task.cpp:1016]: When `noMoreBroadcastBuffers_ == true`, the current logic is ignoring or it is just a no-op, so I feel returning true helps the caller. It is easier for caller to make out if the error case is coming from task being deleted and can take necessary action.

**Comment by karteekmurthys** [comment]
@mbasmanova Addressed comments. Would you please review this.

File: velox/exec/Task.h [hunk 7]
```diff
@@ -194,7 +194,9 @@ class Task : public std::enable_shared_from_this<Task> {
   /// of buffers. No more calls are expected after the call with noMoreBuffers
   /// == true, but occasionally the caller might resend it, so calls
   /// received after a call with noMoreBuffers == true are ignored.
-  void updateBroadcastOutputBuffers(int numBuffers, bool noMoreBuffers);
+  /// @return bool Return true if the update was successful, else false. Also,
+  /// returns true if noMoreBroadcastBuffers_.
```
Comment (mbasmanova) [inline velox/exec/Task.h:198]: noMoreBroadcastBuffers_ is a private member variables; the users of Task may not know how it is set. Consider explaining the API in public terms only: "Returns true without making any changes if a call with noMoreBuffers = true was received earlier."

Perhaps, combine the 2 sentences. "@return true if the update was successful or there was an earlier call with noMoreBuffers == true. Returns false if task failed and buffers were removed."

File: velox/exec/tests/TaskTest.cpp [hunk 8]
```diff
@@ -627,6 +628,34 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
   ASSERT_FALSE(task->supportsSingleThreadedExecution());
 }
 
+TEST_F(TaskTest, updateBroadCastOutputBuffersTest) {
```
Comment (mbasmanova) [inline velox/exec/tests/TaskTest.cpp:631]: naming: drop Test suffix: updateBroadCastOutputBuffers

File: velox/exec/tests/TaskTest.cpp [hunk 9]
```diff
@@ -627,6 +628,34 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
...
+  auto plan = PlanBuilder()
+                  .tableScan(ROW({"c0"}, {BIGINT()}))
+                  .project({"c0 % 10"})
+                  .partitionedOutputBroadcast(std::vector<std::string>{"p0"})
```
Comment (mbasmanova) [inline velox/exec/tests/TaskTest.cpp:635]: nit: drop std::vector<std::string>

File: velox/exec/tests/TaskTest.cpp [hunk 10]
```diff
@@ -627,6 +628,34 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
...
+                  .planFragment();
+  auto taskId = "single.execution.task.0";
+  auto task = std::make_shared<exec::Task>(
+      taskId, plan, 0, std::make_shared<core::QueryCtx>(driverExecutor_.get()));
+  auto bufferManager = PartitionedOutputBufferManager::getInstance();
+  {
+    auto bufferMgrShared = bufferManager.lock();
```
Comment (mbasmanova) [inline velox/exec/tests/TaskTest.cpp:642]: nit: bufferMgrShared -> manager

File: velox/exec/tests/TaskTest.cpp [hunk 11]
```diff
@@ -627,6 +628,34 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
...
+
+  ASSERT_TRUE(task->updateBroadcastOutputBuffers(10, false));
+  {
+    task->requestCancel();
+    auto bufferMgrShared = bufferManager.lock();
+    bufferMgrShared->removeTask(taskId);
+  }
+  ASSERT_FALSE(task->updateBroadcastOutputBuffers(11, true));
+
+  // updateBroadCast will return true. Ignores this broadcast call as previously
```
Comment (mbasmanova) [inline velox/exec/tests/TaskTest.cpp:654]: This API is confusing....

**Review by mbasmanova — APPROVED** [review]
@karteekmurthys Looks fine, but the API is confusing.

File: velox/exec/tests/TaskTest.cpp (hunk 9 above)
Comment (aditi-pandit) [inline velox/exec/tests/TaskTest.cpp:635]: What is p0 here ? Please can you add a comment what it means ?

File: velox/exec/tests/TaskTest.cpp [hunk 12]
```diff
@@ -627,6 +628,34 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
...
+  auto taskId = "single.execution.task.0";
+  auto task = std::make_shared<exec::Task>(
+      taskId, plan, 0, std::make_shared<core::QueryCtx>(driverExecutor_.get()));
+  auto bufferManager = PartitionedOutputBufferManager::getInstance();
+  {
+    auto bufferMgrShared = bufferManager.lock();
+    bufferMgrShared->initializeTask(task, true, 2, 2);
+  }
+
+  ASSERT_TRUE(task->updateBroadcastOutputBuffers(10, false));
```
Comment (aditi-pandit) [inline velox/exec/tests/TaskTest.cpp]: This is a test for noMoreBuffers ? Please state so.

File: velox/exec/tests/TaskTest.cpp [hunk 13]
```diff
@@ -627,6 +628,34 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
...
+  {
+    task->requestCancel();
+    auto bufferMgrShared = bufferManager.lock();
```
Comment (aditi-pandit) [inline velox/exec/tests/TaskTest.cpp:652]: nit : rename "manager"

File: velox/exec/tests/TaskTest.cpp (hunk 11 above)
Comment (aditi-pandit) [reply velox/exec/tests/TaskTest.cpp]: Agree. I can't follow this at all.

File: velox/exec/tests/TaskTest.cpp [hunk 14]
```diff
@@ -627,6 +628,42 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
...
+
+  {
+    // Mock task cancellation.
+    task->requestCancel();
+    auto bufferMgrShared = bufferManager.lock();
+    bufferMgrShared->removeTask(taskId);
+  }
+
+  // Try update after task was cancelled and buffer for that task was removed.
+  ASSERT_EQ(
```
Comment (aditi-pandit) [inline velox/exec/tests/TaskTest.cpp:657]: Was there any reason to change the numBuffers from 10 to 11 ?

File: velox/exec/tests/TaskTest.cpp [hunk 15]
```diff
@@ -627,6 +628,42 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
...
+      task->updateBroadcastOutputBuffers(11, true),
+      UpdateBroadcastStatus::BUFFERS_NOT_FOUND);
+
+  // Returns NO_OP as the previous call set noMoreBuffers to true.
```
Comment (aditi-pandit) [inline velox/exec/tests/TaskTest.cpp:661]: Maybe we should separate the test for BUFFERS_NOT_FOUND from that for NO_OP. Is that possible ?

File: velox/exec/tests/TaskTest.cpp (hunk 14 above)
Comment (karteekmurthys) [reply velox/exec/tests/TaskTest.cpp:657]: This is just a mock up. The numbers don't really mean anything.

File: velox/exec/tests/TaskTest.cpp (hunk 15 above)
Comment (karteekmurthys) [reply velox/exec/tests/TaskTest.cpp]: I have re-ordered the tests and isolated in code-blocks.

File: velox/exec/tests/TaskTest.cpp (hunk 9 above)
Comment (karteekmurthys) [reply velox/exec/tests/TaskTest.cpp:635]: It is the output layout. It is not required to set it. Leaving it empty.

> @karteekmurthys Looks fine, but the API is confusing.

I have improved it by changing the return type to enum. Please take a look.

**Comment by karteekmurthys** [comment]
@mbasmanova I have improved the API to return enum instead of enum.

**Review by aditi-pandit — APPROVED** [review]
Thanks. The changes look good.

File: velox/exec/Task.h [hunk 16]
```diff
@@ -30,7 +30,7 @@ class PartitionedOutputBufferManager;
 
 class HashJoinBridge;
 class CrossJoinBridge;
-
+enum UpdateBroadcastStatus { SUCCESS, NO_OP, BUFFERS_NOT_FOUND };
```
Comment (mbasmanova) [inline velox/exec/Task.h:33]: @karteekmurthys Let's discuss offline the design for this API. 

Please, note that enum values should be named as kSuccess, kNoOp, kBuffersNotFound. Check out the https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md

File: velox/exec/Task.h [hunk 17]
```diff
@@ -194,7 +194,12 @@ class Task : public std::enable_shared_from_this<Task> {
   /// of buffers. No more calls are expected after the call with noMoreBuffers
   /// == true, but occasionally the caller might resend it, so calls
   /// received after a call with noMoreBuffers == true are ignored.
-  void updateBroadcastOutputBuffers(int numBuffers, bool noMoreBuffers);
+  /// @return kSuccess if update was successful.
+  ///         kNoOp if noMoreBuffers was previously set to true.
+  ///         kBuffersNotFound if buffer was not found for a given task.
+  UpdateBroadcastStatus updateBroadcastOutputBuffers(
```
Comment (mbasmanova) [inline velox/exec/Task.h:200]: @karteekmurthys @aditi-pandit 

I took another look at this PR and the corresponding PrestoDB PR. I looks like Task::updateBroadcastOutputBuffers API doesn't need to return anything. This API can continue to be void and just be modified to not throw if task fails while this API is executing. This change won't need any additional changes in Prestissimo.

Notice how similar condition is handled in TaskManager::getDataForResultRequests(resultRequests) in presto-native-execution/presto_cpp/main/TaskManager.cpp. This call doesn't terminate execution of TaskManager::createOrUpdateTask with early return, hence, there is not need for broadcast buffer logic to do so.

Comment (karteekmurthys) [reply velox/exec/Task.h:200]: > Notice how similar condition is handled in TaskManager::getDataForResultRequests(resultRequests) in presto-native-execution/presto_cpp/main/TaskManager.cpp. This call doesn't terminate execution of TaskManager::createOrUpdateTask with early return, hence, there is not need for broadcast buffer logic to do so.

Agree, we may not return early but we do setValue in the `promise` which triggers the `future` that handled here: https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/TaskResource.cpp#L415. This promise and future in ResultRequests are created by this call `TaskManager::getResults()`. 
So we do take action if TaskManager::getDataForResultRequests(resultRequests.

Comment (mbasmanova) [reply velox/exec/Task.h:200]: Is there a similar action we need to take in case of a failure to update broadcast buffers?

Comment (karteekmurthys) [reply velox/exec/Task.h:200]: So I am following what TaskManager::createOrUpdateTask() does if it encountered error:
https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/TaskManager.cpp#L277

Comment (mbasmanova) [reply velox/exec/Task.h:200]: I see. This is not necessary though. Think about what happens if the task fails right after updateBroadcastOutputBuffers returned. The processing in the TaskManager::createOrUpdateTask() will continue as if nothing happened. Hence, it doesn't matter whether task failure during updateBroadcastOutputBuffers triggers early return or not.

Comment (karteekmurthys) [reply velox/exec/Task.h:200]: Alright, I will revert changes to Task::UpdateBroadCastBuffers call. I won't take any action on broadcast failure.

**Comment by karteekmurthys** [comment]
@mbasmanova would you pls help with the merge?

File: velox/exec/tests/TaskTest.cpp (hunk 13 above)
Comment (mbasmanova) [reply velox/exec/tests/TaskTest.cpp:669]: +1

File: velox/exec/tests/TaskTest.cpp [hunk 18]
```diff
@@ -627,6 +628,55 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
...
+    auto task = std::make_shared<exec::Task>(
+        taskId,
+        plan,
+        0,
+        std::make_shared<core::QueryCtx>(driverExecutor_.get()));
+    {
+      auto bufferManagerShared = bufferManager.lock();
+      bufferManagerShared->initializeTask(task, true, 2, 2);
+    }
+    ASSERT_EQ(
```
Comment (mbasmanova) [inline velox/exec/tests/TaskTest.cpp:650]: Use ASSERT_TRUE and ASSERT_FALSE. Update comments to remove references to SUCCESS, NO_OP and BUFFERS_NOT_FOUND enum values that no longer exist.

File: velox/exec/tests/TaskTest.cpp [hunk 19]
```diff
@@ -627,6 +628,55 @@ TEST_F(TaskTest, supportsSingleThreadedExecution) {
...
+
+  // Test for BUFFERS_NOT_FOUND case.
+  {
+    auto taskIdCancelled = "cancelled.execution.task.0";
+    auto taskCancelled = std::make_shared<exec::Task>(
+        taskIdCancelled,
+        plan,
+        0,
+        std::make_shared<core::QueryCtx>(driverExecutor_.get()));
+    auto bufferMgrShared = bufferManager.lock();
```
Comment (mbasmanova) [inline velox/exec/tests/TaskTest.cpp:669]: Can this be simplified to bufferManager.lock()->initializeTask(...)? and bufferManager.lock()->removeTask(...)?

If not, replace bufferMgrShared and bufferManagerShared variables with a single variable bufferManagerShared.

**Review by mbasmanova — COMMENTED** [review]
@karteekmurthys Karteek, the code changes look good. Some comments for the test.

**Comment by karteekmurthys** [comment]
> @karteekmurthys Karteek, the code changes look good. Some comments for the test.

Addressed them. @mbasmanova

**Comment by mbasmanova** [comment]
@karteekmurthys Thank you, Karteek.
